### PR TITLE
continue #15456 add #pragma directives compiler support

### PR DESCRIPTION
**Pull request by ringabout**
After discussing with @juancarlospaco, now I continue PR #15456.

Thank for the efforts from @juancarlospaco!

Related: https://github.com/nim-lang/Nim/issues/1797

```nim
for i in unroll(0, 20, 4):
  echo i
```
Generates C code
```c
#pragma GCC unroll 4
for (i = ((NI) 0); i <= ((NI) 20); ++i)	{
		tyArray__nHXaesL0DJZHyVS07ARPRA T2_;
		nimZeroMem((void*)T2_, sizeof(tyArray__nHXaesL0DJZHyVS07ARPRA));
		T2_[0] = nimIntToStr(i);
		echoBinSafe(T2_, 1);
	}

```

After this PR, it also supports introducing other pragma functions

For example
```nim
iterator vectorization*[S, T](a: S, b: T, annotation: static string = "omp simd"): T {.
  inline, magic: "OmpParFor", sideEffect.} =
  discard
```

- [x] We could change the name of `OmpParFor` in the future PR.

### CODE REVIEW

**Comment by ringabout** [comment]
ASM looks like

```asm
call	nimIntToStr
call	echoBinSafe

call	nimIntToStr
call	echoBinSafe

call	nimIntToStr
call	echoBinSafe

call	nimIntToStr
call	echoBinSafe
```

**Comment by Araq** [comment]
Please add it to a different module, ``std / loopunrolling`` or similar.

File: tests/stdlib/tunroll.nim [hunk 1]
```diff
@@ -0,0 +1,9 @@
+discard """
+  targets: "c cpp js"
+  action: compile
```
Comment (timotheecour) [inline tests/stdlib/tunroll.nim]: can't this test be turned into `action: run` ?

File: tests/stdlib/tunroll.nim [hunk 2]
```diff
@@ -0,0 +1,9 @@
...
+"""
+
+import std/unrolling
```
Comment (timotheecour) [inline tests/stdlib/tunroll.nim]: can you also test in VM? if needed, use the fallback not unrolled implementation in VM.

Comment (ringabout) [reply tests/stdlib/tunroll.nim]: It doesn't work in VM, there is a bug in VM

Comment (ringabout) [reply tests/stdlib/tunroll.nim]:
```
D:\QQPCmgr\Desktop\Nim\play.nim(7, 18) template/generic instantiation of `unroll` from here
D:\QQPCmgr\Desktop\Nim\lib\system\iterators_1.nim(211, 13) Error: attempting to call routine: 'unroll'
  found system.unroll(a: S, b: T, annotation: static[string]) [iterator declared in D:\QQPCmgr\Desktop\Nim\lib\system\iterators_1.nim(173, 10)]
Traceback (most recent call last)
```

Comment (timotheecour) [reply tests/stdlib/tunroll.nim]: use this: (i tested, it works)
```nim
discard """
  targets: "c cpp js"
"""

import std/unrolling
from sugar import collect

template fn() =
  var s: seq[int]
  for i in unroll(0, 100, 5):
    s.add i

  let s2 = collect:
    for i in 0..100: i
  doAssert s == s2

# static: fn() # xxx: Error: cannot generate VM code for for i`gensym0 in pragmaHint(0, 100, "unroll 5"):
fn()
```

Comment (xflywind) [reply tests/stdlib/tunroll.nim:6]: ?

Comment (ringabout) [reply tests/stdlib/tunroll.nim]: I see, thanks

**Comment by Araq** [comment]
Sorry, now that I re-think this... what's the point? Unrolling is trivial to do with a template in Nim and then it doesn't require yet another codegen extension.

**Comment by ringabout** [comment]
> Sorry, now that I re-think this... what's the point? Unrolling is trivial to do with a template in Nim and then it doesn't require yet another codegen extension.

Could we keep the compiler changes? So that it could be implemented as a nimble package.

It is also helpful to #1797

**Comment by Araq** [comment]
Meh, fine.

**Pull request merged**